synapse/logging/context.py

```diff
@@ -778,7 +778,10 @@ def run_in_background(f, *args, **kwargs) -> defer.Deferred:
     return res
 
 
-def make_deferred_yieldable(deferred):
+T = TypeVar("T")
+
+
+def make_deferred_yieldable(deferred: "defer.Deferred[T]") -> "defer.Deferred[T]":
     """Given a deferred (or coroutine), make it follow the Synapse logcontext
     rules:
```

Comment:
Judging by the doctstring I'd've thought the annotation ought to be `Union["defer.Deferred[T]", Awaitable[T]]`.

This might have cropped up in https://github.com/matrix-org/synapse/pull/10980 --- I remember something like this being painful.

Reply:
Oops, I'd entirely forgotten that there was a previous PR that tried to type hint these. I didn't mean to step on your toes!

Yes, the docstring contradicts the type hint. I wasn't sure what to do here. I can't find anywhere where we feed a coroutine to this function and I would remove the coroutine support, but it feels risky?

I could change it to accept an `Awaitable`, then I'd need another assert in the implementation, since we can't really handle non-coroutine, non-deferred awaitables.

Reply:
Not to worry---I had let that one rot a little and I don't think I had as robust an understanding of what's going on then as you do now!

Maybe we can remove `(or coroutine)` from the docstring, if we're convinced that this always is given a `Deferred`.

Reply:
I've elected to remove the coroutine support, since mypy thinks it is unused, and I can't find any usage with coroutines manually.